src/inmanta/deploy/scheduler.py

```diff
@@ -502,6 +504,11 @@ async def _new_version(
                     added_requires[resource] = added
                 if dropped:
                     dropped_requires[resource] = dropped
+                    if self._state.resource_state[resource].blocked is BlockedStatus.TRANSIENT:
+                        dependencies = await self._get_last_non_deploying_state_for_dependencies(resource=resource)
+                        if all(dependencies[requires_id].deployed for requires_id in new_requires):
+                            self._state.resource_state[resource].blocked = BlockedStatus.NO
+                            unblocked_resources.add(resource)
```

Comment:
@sander Here is where we check the dropped requires. I make sure that all the requires are `deployed`, is this the correct check? Should we unblock them if they are in another non failing state (i.e. deploying or available)?

Reply:
Technically I think we want to wait for `ComplianceStatus.COMPLIANT` (not calling `_get_last_non_...()`), but I wonder if we might want to take the easy way out and simply unblock them (a bit too eagerly, which is a freedom we have: it *might* be unblocked, so we give it another shot) if `dropped` is not empty (i.e. at least one dependency was dropped). Main motivation: don't add even more complexity to this method for limited gains. Secondary motivation: performance while under scheduler lock, though this may be negligible (**important side note: we're not under scheduler lock here but we should be if we read from `self._state`, so if we keep the check this must be moved further down**).

I'm honestly not sure. The logic to do it accurately isn't too complex, but is it worth it, given that it might draw the focus from the important parts of this method? Perhaps we can get the best of both worlds if we incorporate it in `State.update_requires()`? But I'd only do it if it doesn't complicate that method too much. There's still the performance aspect but I think it should be acceptable.

What do you think?

Reply:
```
                if dropped:
                    dropped_requires[resource] = dropped
                    if self._state.resource_state[resource].blocked is BlockedStatus.TRANSIENT:
                            self._state.resource_state[resource].blocked = BlockedStatus.NO
```
If we do it like this, aren't we still reading from `self._state` and need to do it under `scheduler_lock`. Maybe we should do it next to update_requires, anyway.
I'm fine either way, leaning more towards just checking if all the dependencies are compliant.

Reply:
Yes, we need to do it under the lock either way, my performance remark was more that the less we do under the lock the faster we can release it again.